include/seastar/rpc/rpc.hh

```diff
@@ -96,7 +96,10 @@ struct resource_limits {
     size_t max_memory = rpc_semaphore::max_counter(); ///< Maximum amount of memory that may be consumed by all requests
     /// Configures isolation for a connection based on its isolation cookie. May throw,
     /// in which case the connection will be terminated.
-    std::function<isolation_config (sstring isolation_cookie)> isolate_connection = default_isolate_connection;
+    using syncronous_isolation_function = std::function<isolation_config (sstring isolation_cookie)>;
+    using asyncronous_isolation_function = std::function<future<isolation_config> (sstring isolation_cookie)>;
+
+    std::variant<syncronous_isolation_function, asyncronous_isolation_function> isolate_connection = default_isolate_connection;
```

Comment:
Good idea to use variant here. Add #include <variant> to be sure we don't lose it if we rearrange headers.

Does it work in C++17 mode? I don't know how good the std::function constructor is at rejecting inputs with the wrong signature.

Reply:
We can also just store an async method here, and wrap synchronous methods. THis is what essentially happens when this variant is used.

Reply:
@denesb I will try to implement your idea. I will convert it to a wrapper class.


> Does it work in C++17 mode? I don't know how good the std::function constructor is at rejecting inputs with the wrong signature.

Honestly, I forgot to check, I will test it after I make the changes suggested by @denesb.

Reply:
@denesb I tried to implement a function wrapper myself and it wasn't trivial for me (to say the least). The best idea I had
was to pull out the variant code I wrote into a dedicated struct that will handle the invocation (and so I did).

> Does it work in C++17 mode? I don't know how good the std::function constructor is at rejecting inputs with the wrong signature.

@avikivity it does work, I tried to remove the sync option and tried to compile, and then 
removed the synchronous option and converted the function (`default_isolate_connection`) to return a future and tried
to compile, in both cases the compilation failed on `no known conversion` on the assignment of the default function.

One more point - maybe I should try and generalize this as a follow up, it sounds a pretty trivial need to have a synchronous and asynchronous versions invoked in a unified manner.